pkg/kubectl/cmd/get/humanreadable.go

```diff
@@ -0,0 +1,874 @@
+	}
+	funcType := printFunc.Type()
+	if funcType.NumIn() != 2 || funcType.NumOut() != 2 {
+		return fmt.Errorf("invalid print handler." +
+			"Must accept 2 parameters and return 2 value.")
+	}
+	if funcType.In(1) != reflect.TypeOf((*PrintOptions)(nil)).Elem() ||
+		funcType.Out(0) != reflect.TypeOf((*[]metav1beta1.TableRow)(nil)).Elem() ||
+		funcType.Out(1) != reflect.TypeOf((*error)(nil)).Elem() {
+		return fmt.Errorf("invalid print handler. The expected signature is: "+
+			"func handler(obj %v, options PrintOptions) ([]metav1beta1.TableRow, error)", funcType.In(0))
+	}
+	return nil
+}
+
+// ValidatePrintHandlerFunc validates print handler signature.
+// printFunc is the function that will be called to print an object.
+// It must be of the following type:
+//  func printFunc(object ObjectType, w io.Writer, options PrintOptions) error
+// where ObjectType is the type of the object that will be printed.
+// DEPRECATED: will be replaced with ValidateRowPrintHandlerFunc
+func ValidatePrintHandlerFunc(printFunc reflect.Value) error {
+	if printFunc.Kind() != reflect.Func {
+		return fmt.Errorf("invalid print handler. %#v is not a function", printFunc)
+	}
+	funcType := printFunc.Type()
+	if funcType.NumIn() != 3 || funcType.NumOut() != 1 {
+		return fmt.Errorf("invalid print handler." +
+			"Must accept 3 parameters and return 1 value.")
+	}
+	if funcType.In(1) != reflect.TypeOf((*io.Writer)(nil)).Elem() ||
+		funcType.In(2) != reflect.TypeOf((*PrintOptions)(nil)).Elem() ||
+		funcType.Out(0) != reflect.TypeOf((*error)(nil)).Elem() {
+		return fmt.Errorf("invalid print handler. The expected signature is: "+
+			"func handler(obj %v, w io.Writer, options PrintOptions) error", funcType.In(0))
+	}
+	return nil
+}
+
+func (h *HumanReadablePrinter) HandledResources() []string {
+	keys := make([]string, 0)
+
+	for k := range h.handlerMap {
+		// k.String looks like "*corev1.PodList" and we want just "pod"
+		api := strings.Split(k.String(), ".")
+		resource := api[len(api)-1]
+		if strings.HasSuffix(resource, "List") {
+			continue
+		}
+		resource = strings.ToLower(resource)
+		keys = append(keys, resource)
+	}
+	return keys
+}
+
+func (h *HumanReadablePrinter) unknown(data []byte, w io.Writer) error {
+	_, err := fmt.Fprintf(w, "Unknown object: %s", string(data))
+	return err
+}
+
+func printHeader(columnNames []string, w io.Writer) error {
+	if _, err := fmt.Fprintf(w, "%s\n", strings.Join(columnNames, "\t")); err != nil {
+		return err
+	}
+	return nil
+}
+
+// PrintObj prints the obj in a human-friendly format according to the type of the obj.
+func (h *HumanReadablePrinter) PrintObj(obj runtime.Object, output io.Writer) error {
+	w, found := output.(*tabwriter.Writer)
+	if !found && !h.skipTabWriter {
+		w = utilprinters.GetNewTabWriter(output)
+		output = w
+		defer w.Flush()
+	}
+
+	// display tables following the rules of options
+	if table, ok := obj.(*metav1beta1.Table); ok {
+		if err := DecorateTable(table, h.options); err != nil {
```

Comment:
is this replicating table-printing code? that part seems like it should move to cli-runtime. I expected the only duplication in this PR to be things we could completely delete after 1.16, once we have support for server-side printing of watch events

Reply:
moving the existing DecorateTable and PrintTable to cli-runtime/pkg/printers/table.go would make sense to me

cc @smarterclayton for plans for collapsing onto server-side printing

Reply:
I believe a significant portions of `humanreadable.go` will be used by both `kubectl` and `registry` after 1.15. In that case, the best way forward may be to just move all of `humanreadable.go` (and `humanreadable_test.go` as well as `PrintOptions`) into a common location out of core: `cli-runtime`. The copied external type handlers (`printer.go`) will be clearly marked as removable after 1.15 and can be moved to a subdirectory of `get` called `legacy` or renamed `legacy_printers.go`. Let me know what you think.

Reply:
moving the bits needed to print tables to cli-runtime makes sense to me

xref https://github.com/kubernetes/kubernetes/issues/74179

https://github.com/kubernetes/kubernetes/pull/76161 allows kubectl to use server-side printing for all get/watch requests

https://github.com/kubernetes/kubernetes/pull/76272 removes unused methods from humanreadable printer and removes internal decoding from it (if decoding to internal is required, it should be done before passing to PrintObj, like kubectl get currently does)